### DISCUSSION TITLE
Internal DNS

### DISCUSSION THREAD

**Discussion by cyrilghali** [onedr0p/cluster-template#719] [Q&A] [answered]
Hello! I did run the task ansible:nuke and I did follow each steps from the readme of the v4.0.1 template.

First, DNS wasn't correctly setuped on cloudlfare.
My old dns settings that was prepared and apply with terraform before was still there.
I did replace each occurrence of the word "ipv4" by "ingress" in DNS Cloudflare settings accordingly to this commit https://github.com/onedr0p/flux-cluster-template/commit/399f68627928b916a0a21b68711e55f175e5d309 and after this I could access to `https://echo-server.${SECRET_DOMAIN}`. I find that very strange because Cloudflare Tunnel does not need any dns settings to work
I'm not aware if anything else needs to be done, or if this is the right thing to do ? Maybe the README could be updated for migration?

My second issue is non related to the migration, but each internal dns request like the kubernetes dashboard or hajimari doesn't work/ are not resolved. My router have one dns server, which is metallb gateway IP that I put on my .config.env. Any help on this would be appreciated!

**Comment by onedr0p** [comment]
> My old dns settings that was prepared and apply with terraform before was still there.

Yes part of the migration to cloudflared would be to remove the ipv4 references from your records and any TXT records as well. But since this CNAME changed to `ingress` it is not required... just something worth cleaning up.

> I find that very strange because Cloudflare Tunnel does not need any dns settings to work

This implementation uses external-dns to populate DNS records that points to your cfargotunnel address. So yes, there does need to be DNS records for cloudflared to work.

> Maybe the README could be updated for migration?

I would rather keep that here in the discussions so people are chime in with their thoughts. 😄 

> My second issue is non related to the migration, but each internal dns request like the kubernetes dashboard or hajimari doesn't work/ are not resolved. My router have one dns server, which is metallb gateway IP that I put on my .config.env. Any help on this would be appreciated!

I might need a little more details here, internal DNS will not work OOTB and requires setup on another device that does DNS for your home network in order to work. However, you can test with your workstations host file. Since what people use for DNS at home varies, it's hard to describe exactly what you need to do as explained [here](https://github.com/onedr0p/flux-cluster-template#-dns).

Edit: I just updated the README on clearer internal DNS things, hopefully it helps.

**Comment by cyrilghali** [comment]
Thanks for taking the time to support and updating the readme!

I did followed the README and setup a pi hole instance.
My router have one dns server which is my pi-hole ip. And I did add a localdns entry like this. But I can't access to hajimari or kubernetes dashboard

![image](https://user-images.githubusercontent.com/40162921/236661602-8b4336eb-73fb-4b6c-8f94-6740c400df0a.png)


As you can see I can ping pi-hole, and my domain is redirected to the gateway lb address, 
![image](https://user-images.githubusercontent.com/40162921/236661721-878b20d3-9976-4ca3-baf3-014d5ed385bd.png)
![image](https://user-images.githubusercontent.com/40162921/236661780-b62c2060-ccdc-43a1-8097-a3095feb6c7c.png)


Ping `hajimari.${SECRET_DOMAIN}` lead to `No address associated with hostname`. I tried to add hajimari to CNAME records in pihole, and I also tried to put the whole subdomain.domain in a A record, but I get the same error

**Reply by onedr0p** [reply]
Be sure to debug using dig too...

```
dig @<pihole-ip> hajimari.${SECRET_DOMAIN}
dig @<k8s-gateway> hajimari.${SECRET_DOMAIN}
```

**Reply by cyrilghali** [reply]
I appreciate your help!

It seems that a both command didn´t return any error. Pi-hole didn't reveive an answer, but k8s-gateway did with a warning. 

The dig with the k8s gateway  answered with my ingress lb ip address, but pi-hole didn't get an answer.

I did search on google about the warning "WARNING: recursion requested but not available"


As you said it won't work out of the box. I'll check some tutorials on how to setup pihole and k8s gateway.

I'll update this discussion if needed

Here is the output

```
$ dig @192.168.1.253 hajimari.cyrilghali.fr

; <<>> DiG 9.18.14 <<>> @192.168.1.253 hajimari.cyrilghali.fr
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 7306
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;hajimari.cyrilghali.fr.		IN	A

;; Query time: 5 msec
;; SERVER: 192.168.1.253#53(192.168.1.253) (UDP)
;; WHEN: Sun May 07 13:22:17 CEST 2023
;; MSG SIZE  rcvd: 51

cghali at fedora-hp-g7 in ~/home-ops on main
$ dig @192.168.1.220 hajimari.cyrilghali.fr

; <<>> DiG 9.18.14 <<>> @192.168.1.220 hajimari.cyrilghali.fr
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 42850
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; COOKIE: bed7b1c45263f539 (echoed)
;; QUESTION SECTION:
;hajimari.cyrilghali.fr.		IN	A

;; ANSWER SECTION:
hajimari.cyrilghali.fr.	1	IN	A	192.168.1.221

;; Query time: 17 msec
;; SERVER: 192.168.1.220#53(192.168.1.220) (UDP)
;; WHEN: Sun May 07 13:22:24 CEST 2023
;; MSG SIZE  rcvd: 101
```

**Reply by joryirving** [reply]
You want the local DNS record on the pihole to be `ingress.${SECRET.DOMAIN}`  rather than the base domain. When I did the migration, I only needed to delete the A record rather than all the DNS records

**Reply by cyrilghali** [reply]
So are you saying that `ingress.${SECRET.DOMAIN}` should point to `BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR` locally?
If yes unfortunately I got the same result!
Even removing any local dns entry gave me the same output log of my two dig command
I'm trying to understand coredns and k8s_gateway but this is not an easy topic for me

**Reply by joryirving** [reply]
That's right. I also had issues where my ISP issued DNS Ipv6 server returned authoritative results over my pihole.

**Reply by cyrilghali** [reply]
I did remove any authoritative upstream dns server like cloudflare on pi-hole and set the dns record just like you said and it worked! I also set the upstream dns server on pi-hole to my gateway address!

Thank you so much to both of you!

**Reply by cyrilghali** [reply, marked as answer]
It seems that just set `BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR` as the only upstream DNS on Pi-hole does the job, without any local dns entry

**Reply by onedr0p** [reply]
I wish there was a more streamlined way to do dns@home but everyone does things a bit differently here due to the vast majority of solutions

**Comment by arthurgeek** [comment]
I use AdGuard Home and when I set DNS rewrite from `*.${SECRET.DOMAIN}` to point to `${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}` it doesn't work. It only works when setting it to `${BOOTSTRAP_METALLB_INGRESS_ADDR}`. Here's `dig` when pointing it to `${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}`:

```
$ dig @${adguard-home-ip} hajimari.${SECRET.DOMAIN}
[...]
;; ANSWER SECTION:
hajimari.${SECRET.DOMAIN}.	10	IN	A	${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}
$ dig @${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR} hajimari.${SECRET.DOMAIN}
[...]
;; ANSWER SECTION:
hajimari.${SECRET.DOMAIN}.	10	IN	A	${BOOTSTRAP_METALLB_INGRESS_ADDR}
```

**Reply by onedr0p** [reply]
What happens when you run: `dig @${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR} hajimari.${SECRET.DOMAIN}` ?

If that returns the address of `${BOOTSTRAP_METALLB_INGRESS_ADDR}` then it's just a matter of getting the correct options in AGH. Basically you need a domain override to tell AGH to use `${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}` when requesting any domain on `${SECRET.DOMAIN}`

**Reply by arthurgeek** [reply]
that is the second dig output in my comment above. it correctly returns the IP of `${BOOTSTRAP_METALLB_INGRESS_ADDR}`.

Not sure which option to set AdGuard Home to. Here's the options:

<img width="487" alt="Screenshot 2023-06-13 at 19 30 13" src="https://github.com/onedr0p/flux-cluster-template/assets/2992/61d01550-5f3d-40a3-bd40-51bb8d286dce">

I'm using the first one: pointing it to `${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}` IP address.

**Reply by onedr0p** [reply]
I don't think that is where you want to set that, see [here](https://github.com/AdguardTeam/AdGuardHome/wiki/Configuration#upstreams-for-domains).

e.g.

<img width="610" alt="image" src="https://github.com/onedr0p/flux-cluster-template/assets/213795/8c7c3251-7327-4932-8069-f0c173caf008">

**Reply by arthurgeek** [reply]
ah! was coming here to say that I found out that this is the right place to configure it. setting it this way works perfectly. thank you!

I used to use the DNS rewrite option for my compose apps running behind nginx-proxy-manager.

**Comment by phekno** [comment]
I have a similar set up to one of the previous commenters (AdGuard Home, and I have my upstream config set up for conditional forwarding), but I'm having a different problem.

I was trying to use Jellyfin, which is "external", but because clients on my network are using AdGuard Home so they're forwarding all requests to `${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}`.  Previously that wasn't an issue, but now `${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}` returns NOTHING.  All requests just time out.  Even if I try one of the "internal" services (e.g. hajimari), I get nothing.  This happened before I migrated to cloudflared, but I have since moved over to that and updated my repo accordingly.  I'm not sure if there's something screwy with my network in general, or if it's something else with the migration?  I don't *think* it's my network, because everything else seems fine (i.e. I can SSH to the host machine, I can make kubectl requests to it just fine, etc.).

EDIT: Something is screwy with my k8s-gateway deployment, because it's saying `Helm rollback failed: release k8s-gateway failed: context deadline exceeded...`.  Looks like I'm digging some more...

**Reply by onedr0p** [reply]
You can try the following

```
flux -n networking reconcile hr k8s-gateway
```

and/or

```
flux -n networking suspend hr k8s-gateway && flux -n networking resume hr k8s-gateway
```

**Reply by phekno** [reply]
I tried the first command, and it just said `install retries exhausted`.  Second command(s) gave me the following:
```
► suspending helmrelease k8s-gateway in networking namespace
✔ helmrelease suspended
► resuming helmrelease k8s-gateway in networking namespace
✔ helmrelease resumed
◎ waiting for HelmRelease reconciliation
✗ client rate limiter Wait returned an error: context deadline exceeded
```

What's weird is that the k8s-gateway pod is running just fine with no errors in it's logs.

**Reply by onedr0p** [reply]
That is really odd... not sure I've seen that before. But you could try commenting out  and adding it back.

Comment out:
https://github.com/onedr0p/flux-cluster-template/blob/8c30a120dae7d24e9ea61b6634d53cbac3be660d/kubernetes/apps/networking/kustomization.yaml#L9

Commit & push & validate it's gone from cluster, then uncomment and Commit & push & validate again.

**Reply by phekno** [reply]
Done...but same thing. :(

**Reply by onedr0p** [reply]
See this https://github.com/fluxcd/flux2/issues/2508#issuecomment-1459767937 so try bouncing all the flux pods.

```
kubectl -n flux-system rollout restart deploy
```

**Reply by phekno** [reply]
No dice.  WTF?

Here's part of the output when I describe the hemlrelease:
```
Status:
  Conditions:
    Last Transition Time:          2023-06-15T18:13:26Z
    Message:                       Reconciliation in progress
    Reason:                        Progressing
    Status:                        Unknown
    Type:                          Ready
  Helm Chart:                      flux-system/networking-k8s-gateway
  Last Attempted Revision:         2.0.4
  Last Attempted Values Checksum:  7cf4aa9f5fffa8ade98f455d52dd7982e957dcb6
  Last Handled Reconcile At:       2023-06-15T13:11:17.999735-05:00
  Last Release Revision:           1
  Observed Generation:             1
Events:
  Type     Reason  Age                  From             Message
  ----     ------  ----                 ----             -------
  Normal   info    23m                  helm-controller  HelmChart 'flux-system/networking-k8s-gateway' is not ready
  Normal   info    8m17s (x3 over 18m)  helm-controller  Helm uninstall succeeded
  Normal   info    8m14s (x4 over 23m)  helm-controller  Helm install has started
  Warning  error   3m13s (x4 over 18m)  helm-controller  Helm install failed: context deadline exceeded

Last Helm logs:
  Warning  error  3m13s (x4 over 18m)  helm-controller  reconciliation failed: Helm install failed: context deadline exceeded
  Normal   info   3m13s                helm-controller  Helm upgrade has started
```

EDIT: I should probably add that I've tried nuking and installing my cluster several times...

**Reply by onedr0p** [reply]
Not related but be careful with that since you might get rate limited by Let's Encrypt.

As for the helm issue that is very strange, here is another related issue: https://github.com/fluxcd/flux2/issues/3974

**Reply by phekno** [reply]
Hadn't thought of that.  Good point.

Yeah...at this point, it's clearly not a DNS issue (which I find very hard to believe, because it's ALWAYS DNS), so you can probably safely just ignore this.

**Reply by onedr0p** [reply]
FWIW I just pushed a commit that will use the staging certificate by default
https://github.com/onedr0p/flux-cluster-template/commit/a489ca1159f0af4e3260ab287c78d7c973772754

**Reply by phekno** [reply]
OK...so I ended up having to reboot my machine because I was dumb.  When it came back up, I see this when I look at all the pods I have running:
```
NAMESPACE      NAME                                                READY   STATUS    RESTARTS      AGE
cert-manager   cert-manager-55f4c7c846-4pw5l                       1/1     Running   1 (42m ago)   27h
cert-manager   cert-manager-cainjector-54d99d698f-fswn5            1/1     Running   1 (42m ago)   27h
cert-manager   cert-manager-webhook-7dfc99497f-qf7x6               1/1     Running   1 (42m ago)   27h
default        cloudnative-pg-576569d8cf-rtjnv                     1/1     Running   1 (42m ago)   27h
default        echo-server-786bdf86d5-nzvc2                        1/1     Running   1 (42m ago)   27h
default        hajimari-6f89cf7b66-8q986                           1/1     Running   1 (42m ago)   27h
default        postgres-1                                          1/1     Running   1 (42m ago)   27h
default        redis-node-0                                        3/3     Running   3 (42m ago)   27h
default        redis-node-1                                        3/3     Running   3 (42m ago)   27h
default        redis-node-2                                        3/3     Running   3 (42m ago)   27h
downloads      prowlarr-0                                          1/1     Running   1 (42m ago)   27h
downloads      radarr-0                                            1/1     Running   1 (42m ago)   27h
downloads      recyclarr-0                                         1/1     Running   1 (42m ago)   27h
downloads      sabnzbd-0                                           1/1     Running   1 (42m ago)   27h
downloads      sonarr-0                                            1/1     Running   1 (42m ago)   27h
flux-system    helm-controller-f9f485f69-k9t5s                     1/1     Running   1 (42m ago)   27h
flux-system    image-automation-controller-ffd4bf697-h7qvn         1/1     Running   1 (42m ago)   27h
flux-system    image-reflector-controller-7c558dc48b-pht7x         1/1     Running   1 (42m ago)   27h
flux-system    kustomize-controller-6547bbb5c-t7grd                1/1     Running   1 (42m ago)   27h
flux-system    notification-controller-54f4c79-7mn59               1/1     Running   1 (42m ago)   27h
flux-system    source-controller-6bd7f4959b-rktm2                  1/1     Running   1 (42m ago)   64m
flux-system    weave-gitops-6fbbbfd9b5-kvztn                       1/1     Running   1 (42m ago)   24h
kube-system    cilium-lc8fn                                        1/1     Running   1 (42m ago)   24h
kube-system    cilium-operator-6866dd7b74-q247l                    1/1     Running   1 (42m ago)   24h
kube-system    coredns-77ccd57875-26m7f                            1/1     Running   0             41m
kube-system    csi-nfs-controller-6986f7cd6f-9lvld                 4/4     Running   4 (42m ago)   27h
kube-system    csi-nfs-node-zvscc                                  3/3     Running   3 (42m ago)   27h
kube-system    external-secrets-cert-controller-655fff7c86-ztnrv   1/1     Running   1 (42m ago)   27h
kube-system    external-secrets-dc45ffbbd-9pbtr                    1/1     Running   1 (42m ago)   27h
kube-system    external-secrets-webhook-5567448f67-5g7jg           1/1     Running   1 (42m ago)   27h
kube-system    hubble-relay-644b98bcb6-bqc2z                       1/1     Running   1 (42m ago)   24h
kube-system    hubble-ui-5d7c6975c7-kpg5j                          2/2     Running   2 (42m ago)   24h
kube-system    kube-vip-k8s-0                                      1/1     Running   8 (42m ago)   27h
kube-system    local-path-provisioner-6457bc66fd-c4r6v             1/1     Running   1 (42m ago)   27h
kube-system    metrics-server-5c5667999-dbfv5                      1/1     Running   1 (42m ago)   27h
kube-system    onepassword-connect-68f88dc48-28lls                 2/2     Running   2 (42m ago)   23h
kube-system    reloader-84898f99cf-vht76                           1/1     Running   1 (42m ago)   27h
kube-system    snapshot-controller-645ff45d7b-lg9kl                1/1     Running   1 (42m ago)   27h
media          jellyfin-0                                          1/1     Running   1 (42m ago)   27h
monitoring     kubernetes-dashboard-7c7798685d-jfbw7               2/2     Running   2 (42m ago)   27h
networking     cloudflared-7677b99fff-x9kmr                        1/1     Running   2 (41m ago)   24h
networking     external-dns-55fd78959-swrz5                        1/1     Running   2 (41m ago)   24h
networking     ingress-nginx-controller-678d49cd6f-drclb           1/1     Running   1 (42m ago)   22h
networking     k8s-gateway-65c589bbf4-2fdqh                        1/1     Running   1 (42m ago)   101m
networking     metallb-controller-5cd9b4944b-bgxcn                 1/1     Running   1 (42m ago)   27h
networking     metallb-speaker-dd4k4                               4/4     Running   5 (41m ago)   27h
security       glauth-c47c5c9dc-tb92x                              1/1     Running   1 (42m ago)   27h
security       glauth-c47c5c9dc-vk7h9                              1/1     Running   1 (42m ago)   27h
```

What's weird (IMO) is that `coredns` has NO restarts, while everyone else has had restarts (from 42m ago, when the machine was restarted).  For giggles I tried resolving stuff, and things seem to be working now.  `k8s-gateway` still says that it's trying to reconcile, but is just kinda sitting there and not doing anything.

**Comment by LeonStoldt** [comment]
Hey 🙂,

I'm relatively new to the world of Kubernetes, flux, and related services. While I'm well-versed in Docker and have already set up a cloudflare and reverse proxy with public services with Docker containers, networking isn't something I deal with regularly. So, configuring things like DNS setup only comes up when necessary and my Pihole and DoH configuration was some time ago. Unfortunately, when it comes to the Kubernetes setup / internal DNS and troubleshooting where requests are getting stuck, I find it quite challenging to keep track.

## Current Situation
Let me outline my current situation and configuration:
Router: IP  `${ROUTER_IP}` e.g. 192.168.0.1
Bind9: IP `${BIND9_IP}` e.g. 192.168.0.100 (Docker container)
PiHole: IP `${PIHOLE_IP}` e.g. 192.168.0.101 (Docker container)
Cloudflared: IP `${CLOUDFLARED_IP}` e.g. 192.168.0.102 (Currently using an existing cloudflared Docker container outside the Kubernetes cluster for DNS-Over-HTTPS)
Kubernetes Cluster: IP `${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}` e.g. 192.168.0.220

## Current Configuration:
### Router
My Router has `${BIND9_IP}` configured as the default DNS for all clients in my local network.

### Bind9
Bind9 is configured like this:
<details>
 <summary>named.conf</summary>

```sh
acl internal {
  192.168.0.0/24; //internal Clients
  172.19.0.0/24; //some random range which seems to come up by my router for push notifications or something like this
  127.0.0.1;
};

options {
  forwarders {
    ${PIHOLE_IP};
  };
  recursion yes; //necessary?
  allow-query { internal; };
};

zone "${SECRET_DOMAIN}" IN {
  type master;
  file "/etc/bind/${SECRET_DOMAIN}.zone";
};
```

</details>

So that it has a zone for ${SECRET_DOMAIN} configured and forwards requests for that zone as I configured in the ${SECRET_DOMAIN}.zone file. Other requests (external) will be forwarded to the PiHole to handle external DNS and blocking/filtering of unwanted requests.

My Zone configuration looks like this (with example entry for Grafana):

<details>
 <summary>${SECRET_DOMAIN}.zone</summary>

```sh
$TTL 2d
$ORIGIN ${SECRET_DOMAIN}.
@       IN      SOA     ns.${SECRET_DOMAIN}.     mail.${SECRET_DOMAIN} (
                        2023063000 ;serial
                        12h ;refresh
                        15m ;retry
                        3w ;expiry
                        2h ;minimum-ttl
                        )
@       IN      NS      ns.${SECRET_DOMAIN}.
ns      IN      A       ${BIND9_IP}
ingress IN      A       ${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}
grafana IN      CNAME   ingress.${SECRET_DOMAIN}.
```

</details>

I also tried `grafana IN      A   ${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}` instead of the cname version. Both did not work for me.

### PiHole
My PiHole Configuration looks like this:
Upstream DNS: ${CLOUDFLARED_IP}#5054 (Cloudflared container) - only

Currently, I don't have activated this setting from the Readme:
```sh
# /etc/dnsmasq.d/99-k8s-gateway-forward.conf
server=/${SECRET_DOMAIN}/${METALLB_K8S_GATEWAY_ADDR}
```
as I want Bind9 to rewrite my domain requests.

## Expected DNS Resolving
My DNS-Chain I would like to achieve (or what I thought would make sense) is:
For internal requests for ${SECRET_DOMAIN} (my zone):
`Client` -> `Bind9` -> `Kubernetes Cluster`

For external requests like e.g., google.com:
`Client` -> `Bind9` -> `PiHole` -> `Cloudflared` -> `Public DNS (e.g., 1.1.1.1)`

I have noticed that something like Conditional Forwarding in Pi-Hole causes issues, so I deactivated this option (was activated before for custom router domain).

When I check the DNS lookup from my client:
<details>
 <summary>nslookup</summary>

```sh
$ nslookup grafana.${SECRET_DOMAIN} ${BIND9_IP}
Server:		${BIND9_IP}
Address:	${BIND9_IP}#53

grafana.${SECRET_DOMAIN}	canonical name = ingress.${SECRET_DOMAIN}.
Name:	ingress.${SECRET_DOMAIN}
Address: ${BOOTSTRAP_METALLB_K8S_GATEWAY_ADDR}
```

</details>

This looks good to me so far (I think). When opening this address in a browser from an internal client, it does not resolve.
My public facing endpoints are reachable (from external network), it's only the internal DNS which is not working right now.

## Questions:
1. I have not fully understood what k8s_gateway does, and I've seen that you put it into the archive in your home-ops repository. Is there a reason for it? Do I still need it, or is it an unnecessary DNS Service in my whole chain?
2. I've seen that you (@onedr0p) have a Bind9 Setup as well. Did I make a mistake configuring something along the way, or does my DNS chain have issues? I just wanted to resolve all the domains like grafana.${SECRET_DOMAIN} (internal available) as well as hajimari.${SECRET_DOMAIN} (external available) in my local network via this Bind9 shortcut and resolve any external Domains with blocking via PiHole. Or does it make sense to have a separate zone for something like home.${SECRET_DOMAIN}? And if so, what would be needed to handle such requests in Kubernetes?
3. My first thought was to put Bind9 and PiHole into the Kubernetes Cluster as well, but this seems to be a bit complicated. Does it make sense to include them in the cluster, or should they stand in front of it?

I have tried a lot of solutions the last days so I don't know what mistake i made that is does not work right now.
I appreciate any help and guidance you can provide. Please feel free to suggest any improvements or corrections to my current setup. Thank you in advance!

---

Update (Jul 2): Without using Bind9, so just `Client` -> `PiHole` and custom DNS Entries like `grafana.${SECRET_DOMAIN} | ${BOOTSTRAP_METALLB_INGRESS_ADDR}` it works. I do not strictly need Bind9, but I thought this would be a convinient way of configuring my custom zone as creating local DNS entries in PiHole is manual.

**Reply by onedr0p** [reply]
Looks like you made some progress but I wanted to answer your questions.

> I have not fully understood what k8s_gateway does, and I've seen that you put it into the archive in your home-ops repository. Is there a reason for it? Do I still need it, or is it an unnecessary DNS Service in my whole chain?

`k8s_gateway` is DNS server that is deployed in your cluster to provide DNS to your ingresses that it auto-discovers. It's meant as a way to provide DNS to your cluster applications without much effort in getting it working. I am not using it anymore and instead have bind9 on my router, external-dns populates bind9 records with my ingress records. You can read more about my DNS setup [here](https://github.com/onedr0p/home-ops#-dns), instead of using a local ad-blocker I pay a low yearly cost to https://nextdns.io for ad-blocking.

> I've seen that you (@onedr0p) have a Bind9 Setup as well. Did I make a mistake configuring something along the way, or does my DNS chain have issues? I just wanted to resolve all the domains like grafana.${SECRET_DOMAIN} (internal available) as well as hajimari.${SECRET_DOMAIN} (external available) in my local network via this Bind9 shortcut and resolve any external Domains with blocking via PiHole. Or does it make sense to have a separate zone for something like home.${SECRET_DOMAIN}? And if so, what would be needed to handle such requests in Kubernetes?

You need to choose which DNS server you want to use to serve records for your cluster, external-dns has support for [pihole](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/pihole.md) (and also [bind9](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/rfc2136.md)) if you choose to drop support for `k8s_gateway` like me.

I don't see why you need bind9 and pihole, you could drop bind9 and go all in on pihole + external-dns which might make things easier for you.

> My first thought was to put Bind9 and PiHole into the Kubernetes Cluster as well, but this seems to be a bit complicated. Does it make sense to include them in the cluster, or should they stand in front of it?

I used to have all DNS for my network ing kubernetes but it didn't pass the WAF as if my cluster goes offline my family cannot reach the internet. I would advise against doing it and instead keep using pihole or bind9 on a external device.

**Reply by LeonStoldt** [reply]
Thanks for your answers! I really appreciate it!

I think I'll give it a try to drop k8s_gateway as well and use external-dns like you did (but for pihole instead).
Currently, my Pihole is only able to receive Ipv4 requests (ipv6 is disabled) - that's why I have Bind9 first (at the moment). But after some thoughts, sleep and your answer, I think I'll remove Bind9 again and go all in on Pihole as I already have a running setup. With external DNS it should be fine deploying Pihole dns records. And I will migrate my Pihole Instance to a node who can handle ipv6 as well.

Again, thank you for your useful information and I'll give it a try and provide an update if I've done the changes 🙂

**Comment by onedr0p** [comment]
For PiHole the following config should work...

> [!TIP]
> Below is how to configure a Pi-hole for split DNS. Other platforms should be similar.
> 1. Apply this file on the Pihole server while substituting the variables
> ```sh
> # /etc/dnsmasq.d/99-k8s-gateway-forward.conf
> server=/${bootstrap_cloudflare.domain}/${bootstrap_cloudflare.gateway_vip}
> ```
> 2. Restart dnsmasq on the server.
> 3. Query an internal-only subdomain from your workstation (any `internal` class ingresses): `dig @${home-dns-server-ip} echo-server-internal.${bootstrap_cloudflare.domain}`. It should resolve to `${bootstrap_cloudflare.ingress_vip}`.

**Reply by parsec** [reply]
on the latest version of pihole I seem to not be able to get this to work. I updated to the latest v6 version today and now none of my split-DNS is resolving... I can technically set this up in UniFi I believe now, but of course my DHCP is set up to pass my PiHole as my DNS server. I guess one fix could be to set up my Dream Machine Pro to be the internal DNS server and then have it forward to the pihole... I just can't figure out why dnsmasq on my pihole is refusing to resolve the split dns entry.

you have any thoughts? I also dropped a message in the networking channel on Home Operations if it's easier to troubleshoot/hash it out there.

EDIT: someone on the discord told me how to re-enable reading the `dnsmasq.d` config files, but now it's saying there's not reply from `k8s-gateway` at all when it calls for DNS, so I'm thinking pihole v6 still did something weird... because it worked just fine before i upgraded pihole.

**Reply by tholinka** [reply]
pihole v6 completely changed this api. See this issue, for instance, for external dns https://github.com/kubernetes-sigs/external-dns/issues/5113.

They also changed the default settings to not import from `/etc/dnsmasq.d`. Run `pihole-FTL --config misc.etc_dnsmasq_d true` to re-enable that.

**Reply by parsec** [reply]
I ended up scrapping using pihole as the main DNS resolver and went the UniFi API + external-dns route, which seems to be working (except for not being able to access my external resources internally for whatever reason, since it just redirects to my UDM Pro login page, but that seems like it's probably outside the scope of this discussion). I'm just having the UDM do DNS forwarding to the pihole instead.
